### ISSUE TITLE
share: private (password protected) shares

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#3606]
Implement privately shared files, where it is necessary to know a token (password) to access them. That token might be part of the URL (e.g. by appending `token=2jfa02j3r2309wmke`) or asked in a form upon visiting the share.

**Comment by stumpc5** [comment]
I suggest to stick with how big file sharing platforms as dropbox or google docs behave. This behaviour is certainly well-thought and -tested, and it is what people expect to see when it comes to filesharing.

My suggestion was, as google docs does it, to attach a url to a file or folder, for example, 

https://docs.google.com/forms/d/e/1FAIpQLSeciKwKl9GAhoSpTI3lMz-iqDR3sOyXF6vOSubhVeTFdQV7Dw/viewform

and share them independent of their name. Anyone knowing this url string can see the file, there is no additional password check (of course, the url serves as a pasword).

**Comment by haraldschilly** [comment]
ok, I haven't thought this through and I think we can close my ticket then, because a (randomly picked) share has an URL like that: `https://share.cocalc.com/share/2c6cffb41ab0f1a3bba33997800029e21049bebf/multichannel.ipynb?viewer=share`

the part in the middle is like the random part in the google docs url you mention. Or am I missing something obvious?

**Comment by williamstein** [comment]
> Or am I missing something obvious?

Yes.   The project_id could leak some other way, e.g., if you share another file fully publicly.  

Your original suggestion and @stumpc5's are technically equivalent (a token or random generated url are basically the same thing).  The new suggestion is NOT because once the project_Id leaks, an attacker could scan for filenames.

I think your token (or equivalently @stumpc5's random url) would be well worth implementing.  We would add a column to the public_paths table that contains this token (or random url), and change some logic in the share server to respect or use this.  The token approach is nice, since it doesn't require making an index on this extra column.

Let's keep this open, because it's often requested, genuinely provides something new and useful from a security point of view, and is relatively easy to implement.

**Comment by williamstein** [comment]
I will plan to implement what @stumpc5 suggested above, and this ticket should stay open.  It would be I think easy to implement this.

**Comment by williamstein** [comment]
Actually, not so easy due to some subtle security implications of access to the token and setting it.
